### ISSUE TITLE
`joint_logprob` yields wrong results due to broadcasting

### ISSUE THREAD

**Issue by ricardoV94** [aesara-devs/aeppl#49]
We probably need to collapse each factor to a scalar to avoid double counting issues like this:

```python
import aesara.tensor.random as atr
from aeppl import joint_logprob
#%%

x_rv = atr.normal(name="x")
y_rv = atr.normal(x_rv, 1, size=2, name="y")

x = x_rv.clone()
y = y_rv.clone()

logprob = joint_logprob([y_rv], {x_rv: x, y_rv: y})

eval1 = logprob.eval({x: 0, y: [0, 0]}).sum()
eval1  # -3.6757541328186907
#%%

x_rv = atr.normal(name="x")
y_rv1 = atr.normal(x_rv, 1, name="y1")
y_rv2 = atr.normal(x_rv, 1, name="y2")

x = x_rv.clone()
y1 = y_rv1.clone()
y2 = y_rv2.clone()

logprob = joint_logprob([y_rv1, y_rv2], {x_rv: x, y_rv1: y1, y_rv2: y2})

eval2 = logprob.eval({x: 0, y1: 0, y2: 0})
eval2  # array(-2.7568156)
```

This is also more easily dealt with if we work with something like #47, where we can decide how to combine terms after we collect them.

**Comment by brandonwillard** [comment]
I'm not sure that those two log-probabilities represent the same things.

**Comment by ricardoV94** [comment]
The second logprop represents the generative graph, while the first would implies there are two independent x variables

**Comment by brandonwillard** [comment]
For the first case, the model graph is given by `model_graph_1 = FunctionGraph(inputs=[x_rv], outputs=[y_rv])`, but in the second case it's given by the `model_graph_2 = FunctionGraph(inputs=[x_rv], outputs=[y_rv1, y_rv2])`.

If we understand the outputs of `model_graph_1` to be the "observed" variables, there's a clear distinction between the two cases in terms of dimensionality: `model_graph_1` "observes" vectors of length two, while `model_graph_2` "observes" two scalars.

Regardless, I understand what you're referring to, and just want to be sure that these details aren't being overlooked, because they might need to be considered when determining a solution for this.

(N.B.: The second case can be converted into the first case using `Z_rv = at.stack([y_rv1, y_rv2])`, and the `joint_logprob` of this should equal the `joint_logprob` of `model_graph_1`.)

**Comment by ricardoV94** [comment]
I don't think the `logprob` should ever correspond to the output of `model_graph_1`.

In any case, #47 revealed a more clear example in one of the existing tests, involving a `dirichlet` + univariate `normal`:
```python
import numpy as np
import aesara.tensor as at
from aeppl import joint_logprob, logprob

x_rv = at.random.dirichlet(np.ones(3), name="x")
y_rv = at.random.normal(x_rv, 1, name="y")

x = x_rv.clone()
y = y_rv.clone()

joint_logp = joint_logprob([y_rv], {y_rv: y, x_rv: x}).eval({x: np.ones(3)/3, y: np.zeros(3)})
joint_logp  # array([-0.28134691, -0.28134691, -0.28134691])

logp_x = logprob(x_rv, x).eval({x: np.ones(3)/3})
logp_x  # array(0.69314718)

normal_rv = at.random.normal(x, 1, name="normal_rv")
normal = normal_rv.clone()
logp_y = logprob(normal_rv, normal).eval({x: np.ones(3)/3, normal: np.zeros(3)})
logp_y  # array([-0.97449409, -0.97449409, -0.97449409])

joint_logp.sum()  # -0.8440407246008489
logp_y.sum() + logp_x.sum()  # -2.230335085720739
```

```python
import pymc3 as pm  # V3

with pm.Model() as m:
    x = pm.Dirichlet("x", np.ones(3), transform=None)
    y = pm.Normal("y", x, 1, shape=3)
    
m.logp({"x": np.ones(3)/3, "y": np.zeros(3)})  # array(-2.23033509)
```

**Comment by brandonwillard** [comment]
> `model_graph_1` observes **one** vector of length two, while `model_graph_2` observes **two** scalars. 

Yes, that is exactly what was stated: the plural on "vectors" does not mean that the model observes more than one vector, but that the types of its observations are a class of vectors.

> I don't think the `joint_logprob` should ever correspond to the output of `model_graph_1`.

What are you referring to?  The value of `joing_logprob` in your example?

**Comment by ricardoV94** [comment]
> What are you referring to? The value of `joing_logprob` in your example?

Yes. I don't think this logprob can ever be considered correct:

```python

x_rv = atr.normal(name="x")
y_rv = atr.normal(x_rv, 1, size=2, name="y")

x = x_rv.clone()
y = y_rv.clone()

logprob = joint_logprob([y_rv], {x_rv: x, y_rv: y})

eval1 = logprob.eval({x: 0, y: [0, 0]}).sum()
eval1  # -3.6757541328186907
```

It's not a question of RV broadcasting, but of the derived logprob of `x` incorrectly being expanded because of the dimensionality of the `y` logprob.

**Comment by ricardoV94** [comment]
The most obvious example is probably this:

```python
x_rv = at.random.normal(name="x")
y_rv = at.random.normal(name="y", size=10)

x = x_rv.clone()
y = y_rv.clone()

logp = joint_logprob([x_rv, y_rv], {x_rv: x, y_rv: y})
logp.eval({x: 0, y: np.zeroes(10)})
```

**Comment by ricardoV94** [comment]
Fixed in #53
